pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -942,3 +979,106 @@ func (c *K8sOrchestrator) ClearFakeAttached(ctx context.Context, volumeID string
 	}
 	return nil
 }
+
+// initVolumeIDToNodesMap performs all the operations required to initialize
+// the volume id to node names map. It also watches for volume attachment add,
+// update & delete operations, and updates the map accordingly.
+func initVolumeIDToNodesMap(ctx context.Context) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("Initializing volumeID to node name map")
+	k8sOrchestratorInstance.volumeIDToNodesMap = &volumeIDToNodesMap{
+		RWMutex: &sync.RWMutex{},
+		items:   make(map[string][]string),
+	}
+
+	// Set up kubernetes resource listener to listen events on volume attachments
+	k8sOrchestratorInstance.informerManager.AddVolumeAttachmentListener(
+		func(obj interface{}) { // Add.
+			volumeAttachmentAdded(obj)
+		},
+		func(oldObj interface{}, newObj interface{}) { // Update.
+			volumeAttachmentUpdated(oldObj, newObj)
+		},
+		func(obj interface{}) { // Delete.
+			volumeAttachmentDeleted(obj)
+		})
+}
+
+// volumeAttachmentAdded adds a new entry or updates an existing entry
+// in the volumeIDToNodeNames map if the volume attachment status is
+// true
+func volumeAttachmentAdded(obj interface{}) {
+	log := logger.GetLogger(context.Background())
+	volAttach, ok := obj.(*storagev1.VolumeAttachment)
+	if volAttach == nil || !ok {
+		log.Warnf("volumeAttachmentAdd: unrecognized object %+v", obj)
+		return
+	}
+
+	log.Debugf("volume attachment add: volume=%v", volAttach)
+	if volAttach.Status.Attached {
+		volumeID := *volAttach.Spec.Source.PersistentVolumeName
+		nodeName := volAttach.Spec.NodeName
+		nodes := k8sOrchestratorInstance.volumeIDToNodesMap.get(volumeID)
+		found := false
+		for _, node := range nodes {
+			if node == nodeName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			nodes = append(nodes, nodeName)
+			log.Debugf("Adding nodeName %v to volumeID %v:%v map", nodeName, volumeID, nodes)
+			k8sOrchestratorInstance.volumeIDToNodesMap.add(volumeID, nodes)
+		}
+	}
+}
+
+func volumeAttachmentUpdated(oldObj interface{}, newObj interface{}) {
+}
+
+// volumeAttachmentDeleted deletes an entry or removes node name form an
+// existing entry in the volumeIDToNodeNames map if the volume attachment
+// status is false
+func volumeAttachmentDeleted(obj interface{}) {
+	log := logger.GetLogger(context.Background())
+	volAttach, ok := obj.(*storagev1.VolumeAttachment)
+	if volAttach == nil || !ok {
+		log.Warnf("volumeAttachmentDeleted: unrecognized object %+v", obj)
+		return
+	}
+	if !volAttach.Status.Attached {
+		log.Debugf("volume attachment deleted: volume=%v", volAttach)
+		volumeID := *volAttach.Spec.Source.PersistentVolumeName
+		nodeName := volAttach.Spec.NodeName
+		nodes := k8sOrchestratorInstance.volumeIDToNodesMap.get(volumeID)
+		found := false
+		for i, node := range nodes {
+			if node == nodeName {
+				nodes[i], nodes[len(nodes)-1] = nodes[len(nodes)-1], nodes[i]
```

Comment:
Why are we shifting the index from last to current ?
Why not remove the item in contention like this:
`nodes = append(nodes[:i], nodes[i+1:]...)`
I think this is simpler and retains the order as well

Reply:
It was just to remove the item i. Replaced it with above logic.